psi4/driver/procrouting/diis.py

```diff
@@ -181,33 +232,170 @@ def extrapolate(self, *args):
+                jac = self.adiis_gradient, tol=1e-6, options={"maxiter": 200})
+
+        if not result.success:
+            raise Exception("ADIIS minimization failed. File a bug, and include your entire input and output files.")
+
+        return normalize_input(result.x)
+
+    def adiis_populate(self):
+        # We are currently assuming that all of dD and dF fit in-core.
+        # These quantities are N^2, so this should be fine in most cases.
+
+        num_entries = len(self.stored_vectors)
+        dD = [[] for x in range(num_entries)]
+        dF = [[] for x in range(num_entries)]
+        for name, array in zip(["densities", "target"], [dD, dF]):
+            for item_num in range(len(self.template[name])):
+                latest_entry = self.load_quantity(name, len(self.stored_vectors) - 1, item_num)
+                for entry_num in range(num_entries):
+                    temp = self.load_quantity(name, entry_num, item_num, force_new=True)
+                    temp.subtract(latest_entry)
+                    array[entry_num].append(temp)
+
+        self.adiis_linear = np.zeros((num_entries))
+        latest_fock = []
+        for item_num in range(len(self.template["target"])):
+            latest_fock.append(self.load_quantity("target", len(self.stored_vectors) - 1, item_num))
+        for i in range(num_entries):
+            self.adiis_linear[i] = sum(d.vector_dot(f) for d, f in zip(dD[i], latest_fock))
+
+        self.adiis_quadratic = np.zeros((num_entries, num_entries))
+        for i, j in product(range(num_entries), repeat = 2):
+            self.adiis_quadratic[i][j] = sum(d.vector_dot(f) for d, f in zip(dD[i], dF[j]))
+
+        if self.closed_shell:
+            self.adiis_linear *= 2
+            self.adiis_quadratic *= 2
+
+    def ediis_energy(self, x):
+        x = normalize_input(x)
+        ediis_linear = np.array([entry["energy"][0] for entry in self.stored_vectors])
+        return np.dot(ediis_linear, x) + np.einsum("i,ij,j->", x, self.ediis_quadratic, x) / 2
+
+    def ediis_gradient(self, x):
+        c = normalize_input(x)
+        ediis_linear = np.array([entry["energy"][0] for entry in self.stored_vectors])
+        dedc = ediis_linear + np.einsum("i,ij->j", c, self.ediis_quadratic)
+
+        norm_sq = (x**2).sum()
+        dcdx = np.diag(x) * norm_sq - np.einsum("i,j->ij", x ** 2, x)
+        dcdx *= 2 / norm_sq**2
+
+        return np.einsum("i,ij->j", dedc, dcdx)
+
+    def ediis_coefficients(self):
+        from scipy.optimize import minimize
+        self.ediis_populate()
+        result = minimize(self.ediis_energy, np.ones(len(self.stored_vectors)), method="BFGS",
+                jac=self.ediis_gradient, tol=1e-6, options={"maxiter": 200})
+
+        if not result.success:
+            raise Exception("EDIIS minimization failed. File a bug, and include your entire input and output files.")
+
+        return normalize_input(result.x)
+
+    def ediis_populate(self):
```

Comment:
Docstring might be good to add even though function name is pretty self-explanatory